Source/Game/Game/ECS/Systems/TransformSystem.h

```diff
@@ -0,0 +1,291 @@
+		void AddOffset(entt::entity entity, const vec3& offset);
+
+		//api with transform component and entity ID to save lookup
+		void SetPosition(entt::entity entity, ECS::Components::Transform& transform, const vec3 &newPosition);
+		void SetRotation(entt::entity entity, ECS::Components::Transform& transform, const quat &newRotation);
+		void SetScale(entt::entity entity, ECS::Components::Transform& transform, const vec3 &newScale);
+		void SetPositionAndRotation(entt::entity entity, ECS::Components::Transform& transform, const vec3 &newpos, const quat &newrotation);
+		void SetComponents(entt::entity entity, ECS::Components::Transform& transform, const vec3& newpos,const quat &newrotation, const vec3 &newscale);
+		void AddOffset(entt::entity entity, ECS::Components::Transform& transform, const vec3& offset);
+
+		void ParentEntityTo(entt::entity parent, entt::entity child);
+
+		template<typename F>
+		void ProcessMovedEntities(F&& fn)
+		{
+			TransformQueueItem item;
+			while (elements.try_dequeue(item))
+			{
+				fn(item.et);
+			}
+		}
+
+	private:
+		entt::registry* owner;
+		
+		friend struct ECS::Components::Transform;
+		struct TransformQueueItem
+		{
+		public:
+			entt::entity et;
+		};
+
+		moodycamel::ConcurrentQueue<TransformQueueItem> elements;
+	};
+}
+
+namespace ECS::Components
+{
+	struct DirtyTransform
+	{
+	public:
+		u64 dirtyFrame = 0;
+	};
+
+	struct Transform
+	{
+		friend struct ECS::Singletons::TransformSystem;
+
+	public:
+		// We are using Unitys Right Handed coordinate system
+		// +X = right
+		// +Y = up
+		// +Z = forward
+		static const vec3 WORLD_FORWARD;
+		static const vec3 WORLD_RIGHT;
+		static const vec3 WORLD_UP;
+
+		//makes the component use pointer stable references in entt. do not remove
+		static constexpr auto in_place_delete = true;
+
+		vec3 GetLocalForward() const
+		{
+			return glm::toMat4(rotation) * vec4(WORLD_FORWARD, 0);
+		}
+
+		vec3 GetLocalRight() const
+		{
+			return glm::toMat4(rotation) * vec4(WORLD_RIGHT, 0);
+		}
+
+		vec3 GetLocalUp() const
+		{
+			return glm::toMat4(rotation) * vec4(WORLD_UP, 0);
+		}
+
+		mat4a GetMatrix() const;
+
+		mat4a GetLocalMatrix() const
+		{
+			return Math::AffineMatrix::TransformMatrix(position,rotation,scale);
+		}
+
+		void SetDirty(ECS::Singletons::TransformSystem& dirtyQueue, entt::entity ownerEntity)
+		{
+			if (ownerEntity != entt::null)
+			{
+				dirtyQueue.elements.enqueue({ ownerEntity });
+			}
+		}
+
+		const vec3& GetPosition() const {
+			return position;
+		}
+
+		const vec3 GetWorldPosition() const;
+
+		const quat& GetRotation() const {
+			return rotation;
+		}
+		const vec3& GetScale() const {
+			return scale;
+		}
+
+		struct SceneNode* ownerNode;
+
+	private:
+		vec3 position = vec3(0.0f, 0.0f, 0.0f);
+		quat rotation = quat(1.0f, 0.0f, 0.0f, 0.0f);
+		vec3 scale = vec3(1.0f, 1.0f, 1.0f);
+	};
+
+	struct SceneNode {
+		mat4a matrix;
```

Comment:
Missing visibility modifier (Please fix it in all instances where it is missing)